### ISSUE TITLE
Tools for saving and loading decoders

### ISSUE THREAD

**Issue by tcstewar** [nengo/nengo_extras#35]

**Comment by tcstewar** [comment]
Oh, and one of the things I'm thinking about proposing for ```nengo_gui``` is some function hooks to give you access to the simulator object, so you could do something like this at the end of your code in ```nengo_gui```:

```python
def on_sim_done(sim): 
    # this will get triggered when the simulation is over
    weight_saver.save(sim)
```

**Comment by hunse** [comment]
> Use case 2 is even more problematic -- there's currently no way to seed the start of a learning rule with anything other than the result of some Solver.

Sure you can. You just make a connection from the neurons to the post object as in case 1. E.g.
```python
import matplotlib.pyplot as plt
import numpy as np

import nengo

n_neurons = 200
initial = np.random.uniform(-0.0001, 0.0001, size=(1, n_neurons))

with nengo.Network() as model:
    x = nengo.Node(nengo.processes.WhiteSignal(period=100, high=5))
    ystar = nengo.Node(lambda t, x: x**2, size_in=1)
    nengo.Connection(x, ystar)

    a = nengo.Ensemble(n_neurons, 1)
    y = nengo.Node(size_in=1)

    error = nengo.Node(size_in=1)
    nengo.Connection(y, error)
    nengo.Connection(ystar, error, transform=-1)

    conn = nengo.Connection(a.neurons, y, transform=initial,
                            learning_rule_type=nengo.PES(1e-3))
    nengo.Connection(error, conn.learning_rule)

    xp = nengo.Probe(x, synapse=0.01)
    ystarp = nengo.Probe(ystar, synapse=0.01)
    yp = nengo.Probe(y, synapse=0.01)

with nengo.Simulator(model) as sim:
    sim.run(15.0)

plt.plot(sim.trange(), sim.data[xp])
plt.plot(sim.trange(), sim.data[ystarp])
plt.plot(sim.trange(), sim.data[yp])
plt.show()
```

**Comment by tcstewar** [comment]
> Sure you can. You just make a connection from the neurons to the post object as in case 1

Huh.  I had no idea you could do that.  That makes complete sense in hind-sight....

**Comment by HuangBo-Terraloupe** [comment]
Hello tcstewar and hunse,
Thanks for your work, I have few questions and hope for your answer.
The weights of communication channel just np.array, we can save it after the simulation (training).
My questions are:
1, how to fine-tuning the network if we want to train the network based on the pre-training weights? 
2, how to use these weights to do single prediction?

For the second question, it seems like we directly define a function, that do matrix Multiplication, and times the input with the weights.
If it is ok, please show me how to do it in the mnist example:


import nengo
import numpy as np

from keras.datasets import mnist
from keras.utils import np_utils
from vision import Gabor, Mask


class Q_network:

    def __init__(self, input_shape, output_shape, nb_hidden, weights_file):
        '''
        :param input_shape: the input dimension without batch_size
        :param output_shape: the output dimension without batch_size
        :param nb_hidden: the number of neurons in ensemble
        :param weights_file: the path to save weights of connection channel
        '''
        self.input_shape = input_shape
        self.output_shape = output_shape
        self.nb_hidden = nb_hidden
        self.weights_file = weights_file

    def encoder_initialization(self, way="default"):
        if way=="random":
            encoders = np.random.normal(0, 1, size=(self.nb_hidden, self.input_shape))
        else:
            rng = np.random.RandomState(self.output_shape)
            encoders = Gabor().generate(self.nb_hidden, (self.output_shape, self.output_shape), rng=rng)
            encoders = Mask((self.input_shape**0.5, self.input_shape**0.5)).populate(encoders, rng=rng, flatten=True)
        return encoders

    def train_network(self, train_data, train_targets):

        encoders = self.encoder_initialization()
        solver = nengo.solvers.LstsqL2(reg=0.01)

        model = nengo.Network(seed=3)
        with model:
            input = nengo.Ensemble(n_neurons=self.nb_hidden,
                                   dimensions=self.input_shape,
                                   neuron_type=nengo.LIFRate(),
                                   intercepts=nengo.dists.Choice([-0.5]),
                                   max_rates=nengo.dists.Choice([100]),
                                   eval_points=train_data,
                                   encoders=encoders,
                                   )
            output = nengo.Node(size_in=self.output_shape)
            conn = nengo.Connection(input, output,
                                    synapse=None,
                                    eval_points=train_data,
                                    function=train_targets,
                                    solver=solver
                                    )

        with nengo.Simulator(model) as sim:
            _, acts = nengo.utils.ensemble.tuning_curves(input, sim, inputs=train_data)
            weights = sim.data[conn].weights.T
            #result = np.dot(acts, weights)
            np.save("weights.npy", weights)


    def prediction(self):

**Comment by Seanny123** [comment]
Hey @HuangBo-Terraloupe,

Did you end asking your question in [the Nengo forum](https://forum.nengo.ai/t/supervised-learning-with-nengo-mnist-example-focus-on-reload-weights-and-simulator/330?u=seanny123)? If yes, would you mind deleting this comment from this thread?

**Comment by HuangBo-Terraloupe** [comment]
Hello,

Can I leave that for a few days.
and I will delete the comment.

thanks

bo
